### PR TITLE
Add BitShift-28 with signed integer type support

### LINKED ISSUE #2406: Add Support for Additional Input Types for BitShift
BitShift only supports unsigned types in opset 11.
PyTorch supports BitShift operations on even other types (int8, int64, int32, float32, float64), and the only unsigned type in PyTorch is uint8; so the use of the Bitshift operator in PyTorch is currently restricted to this type.
We would like to add support to additional input types for BitShift to support more cases in the PyTorch exporter, without the use of a subgraph.

### PR DESCRIPTION
**Pull request by mcollinswisc**
## Summary

`BitShift` currently only accepts unsigned integer types. This adds signed integer types in a new opset-28 schema.

Fixes https://github.com/onnx/onnx/issues/2406

Will be necessary for https://github.com/onnx/onnx/issues/5497 too.

## Motivation

`BitShift` is the only bitwise operator still restricted to unsigned types. `BitwiseAnd`, `BitwiseOr`, `BitwiseXor` and `BitwiseNot` already support signed types.

a4f44a2 added `BitShift` to support hashing algorithms, which only need unsigned types. But shifting with sign extension may be used in other contexts. See for reference:

* Appendix B of https://arxiv.org/abs/1712.05877
* gemmlowp:
  https://github.com/google/gemmlowp/blob/16e8662c34917be0065110bfcd9cc27d30f52fdf/fixedpoint/fixedpoint.h#L365-L379
* TOSA [ARITHMETIC_RIGHT_SHIFT](https://www.mlplatform.org/tosa/tosa_spec_1_0_0.html#_arithmetic_right_shift) and [RESCALE](https://www.mlplatform.org/tosa/tosa_spec_1_0_0.html#_rescale) definition.
* https://arxiv.org/abs/1905.13298

Shifting of signed integer types is available in other frameworks. This reference implementation continues to use NumPy, and that matches PyTorch and JAX too:

https://gist.github.com/mcollinswisc/7cc504381d6c30ee50ef72745f47c6e5

(PyTorch bit shift supports the signed dtypes and `uint8`, but *not* `uint16`, `uint32` or `uint64`.) MindSpore and Tensorflow are found to produce different results, but only on shift counts that are out of range.

Casting to an unsigned type is not a workaround. For a right shift it turns a (sign-extending) arithmetic shift into a logical shift, which gives a different result for negative input.

## Op Semantics

Semantics are chosen to match NumPy (which jax and PyTorch match too).

Some of the behavior (namely for shifts Y that are out of range: negative or larger than the bit width of X) is *not* fully documented by NumPy, but this new BitShift spec does describe it. The results of NumPy can be explained entirely from their implementation and well-defined C behavior. The NumPy PR https://github.com/numpy/numpy/pull/13739 ensured this:

1. https://github.com/numpy/numpy/pull/13739 added logic that explicitly handles shift amounts Y that are `>=` to the bit width of X
2. While [comments in NumPy](https://github.com/numpy/numpy/blob/5e1d03ffac5f2c0a9c39bfcaa9fc853b2b83151e/numpy/_core/src/umath/loops_autovec.dispatch.c.src#L22-L23) still say handling of negative shifts is TODO, the `size_t` casts at https://github.com/numpy/numpy/blob/5e1d03ffac5f2c0a9c39bfcaa9fc853b2b83151e/numpy/_core/src/npymath/npy_math_internal.h.src#L687 and https://github.com/numpy/numpy/blob/5e1d03ffac5f2c0a9c39bfcaa9fc853b2b83151e/numpy/_core/src/npymath/npy_math_internal.h.src#L697 do deterministically define the behavior: because this cast is defined by [C11 6.3.1.3p2](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) (and continues to be defined in the same way through C23). Negative Y becomes a very large `size_t` value in the `if` condition & the result will be the saturated 0 or -1.

The reference implementation & expected test results continue to simply use NumPy. Results are verified more thoroughly against numpy, torch, and JAX in:

https://gist.github.com/mcollinswisc/7cc504381d6c30ee50ef72745f47c6e5

The opset-28 definition is a superset of opset-11, so version upgrade is trivial.

### CODE REVIEW

File: docs/Changelog.md [hunk 1]
```diff
@@ -33087,6 +33087,62 @@ This version of the operator has been available since version 27 of the default
...
+  Bitwise shift operator performs element-wise operation. For each input element, if the
+  attribute "direction" is "RIGHT", this operator moves its binary representation toward
+  the right side so that the input value is effectively decreased. If the attribute "direction"
+  is "LEFT", bits of binary representation moves toward the left side, which results the
+  increase of its actual value. The input X is the tensor to be shifted and another input
+  Y specifies the amounts of shifting. For example, if "direction" is "Right", X is [1, 4],
+  and S is [1, 1], the corresponding output Z would be [0, 2]. If "direction" is "LEFT" with
+  X=[1, 2] and S=[1, 2], the corresponding output Y would be [2, 8].
+
+  The shift is a fixed-width two's-complement operation that behaves like
```
Comment (justinchuby) [inline docs/Changelog.md]: We are moving the specs to be self contained. Could you update this so that it does not rely on the numpy reference?

**Review by justinchuby — COMMENTED** [review]
LGTM in general. Some comments for the spec. Thanks!

File: docs/Changelog.md (hunk 1 above)
Comment (mcollinswisc) [reply docs/Changelog.md]: Okay, this is done: the spec describes the signed behavior w/o referencing numpy. Though, the description is intended to still exactly match numpy's C implementation (at least currently, more details added to this PR's description).

In passing, I also deleted the reference to Numpy-style broadcasting that was in `BitShift_ver28_doc`: it was mostly redundant anyway because the spec includes `GenerateBroadcastingDocMul()`.

Thanks!

File: onnx/defs/logical/defs.cc [hunk 2]
```diff
@@ -144,25 +144,31 @@ ONNX_OPERATOR_SET_SCHEMA(
...
 Bitwise shift operator performs element-wise operation. For each input element, if the
 attribute "direction" is "RIGHT", this operator moves its binary representation toward
-the right side so that the input value is effectively decreased. If the attribute "direction"
-is "LEFT", bits of binary representation moves toward the left side, which results the
-increase of its actual value. The input X is the tensor to be shifted and another input
+the right side. If the attribute "direction" is "LEFT", bits of binary representation
+moves toward the left side. The input X is the tensor to be shifted and another input
 Y specifies the amounts of shifting. For example, if "direction" is "Right", X is [1, 4],
 and S is [1, 1], the corresponding output Z would be [0, 2]. If "direction" is "LEFT" with
 X=[1, 2] and S=[1, 2], the corresponding output Y would be [2, 8].
```
[the remark this answers is not in the document]
Comment (mcollinswisc) [reply onnx/defs/logical/defs.cc]: Addressed by https://github.com/onnx/onnx/pull/8270/changes/69177b02ddf4c632c1f9780ec4d4f101b584dcd6

**Comment by justinchuby** [comment]
I found two correctness issues that should be addressed before merging:

1. **Negative shift counts rely on unspecified NumPy behavior.** BitShift-28 defines deterministic results for negative shift counts, but `onnx/reference/ops/op_bitshift.py` still delegates directly to `np.left_shift` / `np.right_shift`. NumPy documents shift counts as non-negative, so the current observed behavior is not guaranteed and may change to an error. The backend tests also compute their expected values through the same NumPy calls, so they cannot detect a divergence. Please implement the negative-count branch explicitly in the reference op and use literal expected arrays for these cases.

2. **The 28-to-27 signed-type restriction can be bypassed through the public C++ converter API.** `ImportModelProto` leaves ordinary intermediate node values as `UNDEFINED`, and C++ `ConvertVersion` does not run shape inference. Consequently, a graph such as `Cast(to=INT32) -> BitShift -> Cast`, without intermediate `value_info`, can evade `TypeRestriction` and be converted to an invalid opset-27 signed BitShift. The Python binding's pre-conversion shape inference masks this path. Please make the converter reliably resolve or reject unknown types here and add C++ API coverage using an intermediate signed value without `value_info`.

File: onnx/defs/logical/old.cc [hunk 3]
```diff
@@ -262,4 +262,54 @@
...
+and S is [1, 1], the corresponding output Z would be [0, 2]. If "direction" is "LEFT" with
+X=[1, 2] and S=[1, 2], the corresponding output Y would be [2, 8].
+
+Because this operator supports Numpy-style broadcasting, X's and Y's shapes are
+not necessarily identical.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    BitShift,
+    11,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnx/defs/logical/old.cc:268]: This is the macro-generated debug registration counter from `ONNX_OPERATOR_SET_SCHEMA`; its initializer increments `DbgOperatorSetTracker`, while the variable itself is intentionally not read and is marked `[[maybe_unused]]` in `schema.h`. No operator-specific unused state is introduced here, so this alert is a false positive.

**Comment by justinchuby** [comment]
Addressed both correctness issues in 5ec89958df948dc31c23fa78e94aa078a50975f9:

- The reference evaluator now masks negative and `>= bit width` counts before calling NumPy, produces the BitShift-28 fill result explicitly, and performs signed left shifts through the corresponding unsigned representation so discarded high bits are deterministic. Backend out-of-range expectations are literal rather than computed by the same NumPy shift path. Regression coverage exercises LEFT/RIGHT, broadcasting, extreme negative counts, width-boundary counts, and int8/int16/int32/int64.
- Public C++ `ConvertVersion` now runs shape inference before importing the IR, so signed intermediate values without `value_info` reach `TypeRestriction`. Shape inference was removed from the Python-only wrapper to keep a single shared path. A direct C++ API test verifies `Cast(INT32) -> BitShift -> Cast` is rejected for 28→27 while the equivalent UINT32 graph succeeds.

Validated with `pixi run install`, targeted reference/backend/schema/shape/version-converter pytest runs, the direct `VersionConverterTest.*` gtests, `pixi run gen-all`, PR-file lintrunner with zero errors, and full `pixi run lint`.

**Review by justinchuby — APPROVED** [review]
